CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+<sup>[[link](#def-self-class-methods)]</sup>
+
+  ```Ruby
+  class TestClass
+    # bad
+    def TestClass.some_method
+      # body omitted
+    end
+
+    # good
+    def self.some_other_method
+      # body omitted
+    end
+
+    # Also possible and convenient when you
+    # have to define many class methods.
+    class << self
+      def first_method
+        # body omitted
+      end
+
+      def second_method_etc
+        # body omitted
+      end
+    end
+  end
+  ```
+
+* <a name="alias-method-lexically"></a>
+  Prefer `alias` when aliasing methods in lexical class scope as the
+  resolution of `self` in this context is also lexical, and it communicates
+  clearly to the user that the indirection of your alias will not be altered
+  at runtime or by any subclass unless made explicit.
+<sup>[[link](#alias-method-lexically)]</sup>
+
+  ```Ruby
+  class Westerner
+    def first_name
+      @names.first
+    end
+
+    alias given_name first_name
+  end
+  ```
+
+  Since `alias`, like `def`, is a keyword, prefer bareword arguments over
+  symbols or strings. In other words, do `alias foo bar`, not
+  `alias :foo :bar`.
+
+  Also be aware of how Ruby handles aliases and inheritance: an alias
+  references the method that was resolved at the time the alias was defined;
+  it is not dispatched dynamically.
+
+  ```Ruby
+  class Fugitive < Westerner
+    def first_name
+      'Nobody'
+    end
+  end
+  ```
+
+  In this example, `Fugitive#given_name` would still call the original
+  `Westerner#first_name` method, not `Fugitive#first_name`. To override the
+  behavior of `Fugitive#given_name` as well, you'd have to redefine it in the
+  derived class.
+
+  ```Ruby
+  class Fugitive < Westerner
+    def first_name
+      'Nobody'
+    end
+
+    alias given_name first_name
+  end
+  ```
+
+* <a name="alias-method"></a>
+  Always use `alias_method` when aliasing methods of modules, classes, or
+  singleton classes at runtime, as the lexical scope of `alias` leads to
+  unpredictability in these cases.
+<sup>[[link](#alias-method)]</sup>
+
+  ```Ruby
+  module Mononymous
+    def self.included(other)
+      other.class_eval { alias_method :full_name, :given_name }
+    end
+  end
+
+  class Sting < Westerner
+    include Mononymous
+  end
+  ```
+
+## Exceptions
+
+* <a name="fail-method"></a>
+  Signal exceptions using the `fail` method. Use `raise` only when catching an
+  exception and re-raising it (because here you're not failing, but explicitly
+  and purposefully raising an exception).
+<sup>[[link](#fail-method)]</sup>
+
+  ```Ruby
+  begin
+    fail 'Oops'
+  rescue => error
+    raise if error.message != 'Oops'
+  end
+  ```
+
+* <a name="no-explicit-runtimeerror"></a>
+  Don't specify `RuntimeError` explicitly in the two argument version of
+  `fail/raise`.
+<sup>[[link](#no-explicit-runtimeerror)]</sup>
+
+  ```Ruby
+  # bad
+  fail RuntimeError, 'message'
+
+  # good - signals a RuntimeError by default
+  fail 'message'
```

Comment:
Recommend never using a string like this. We should always have an exception class w/ context. Strings are harder to test and manage than classes specifically because you can rescue a class directly -

`rescue SomeSpecificException ...`

vs

`rescue => e; if e.message == "some specific exception" ...`